drivers/storage/openstack/tests/openstack_test.go

```diff
@@ -0,0 +1,534 @@
+// +build !libstorage_storage_driver libstorage_storage_driver_openstack
+
+package openstack
+
+import (
+	"bytes"
+	"os"
+	"strconv"
+	"strings"
+	"testing"
+
+	log "github.com/Sirupsen/logrus"
+	gofig "github.com/akutz/gofig/types"
+	"github.com/stretchr/testify/assert"
+
+	"github.com/codedellemc/libstorage/api/context"
+	"github.com/codedellemc/libstorage/api/registry"
+	"github.com/codedellemc/libstorage/api/server"
+	apitests "github.com/codedellemc/libstorage/api/tests"
+	"github.com/codedellemc/libstorage/api/types"
+	"github.com/codedellemc/libstorage/api/utils"
+
+	// load the driver
+	openstack "github.com/codedellemc/libstorage/drivers/storage/openstack"
+	openstackx "github.com/codedellemc/libstorage/drivers/storage/openstack/executor"
+)
+
+// configYAML is an embedded config file that specifies the bare minimum
+// configuration for creating an openstack volume.
+var configYAML = []byte(`
+openstack:
+  authURL: ` + os.Getenv("OS_AUTH_URL") + `
+  username: ` + os.Getenv("OS_USERNAME") + `
+  tenantName: ` + os.Getenv("OS_TENANT_NAME") + `
+  domainName: ` + os.Getenv("OS_USER_DOMAIN_NAME") + `
+  password: ` + os.Getenv("OS_PASSWORD") + `
+  availabilityZoneName: ` + os.Getenv("OS_AVAILABILITY_ZONE") + `
+  regionName: ` + os.Getenv("OS_REGION_NAME") + `
```

Comment:
I've been trying to think about this so that I don't steer you wrong. The more I've looked at the existing tests, the more I realize that most of their _built-in_ configs are totally unnecessary. It's really the first I've looked at it this closely. Aren't you the lucky one?

> I can just create an empty config byte array and apitests.Run will fill it for me ?

Yes you can. in fact it can be `nil`, as seen in the [`vbox`](https://github.com/codedellemc/libstorage/blob/master/drivers/storage/vbox/tests/vbox_test.go#L95) driver. The tests get compiled into a binary, and when that binary executes, it will automatically look for the env vars defined by the driver's config registration.

Taking EBS as an example again, The [README](https://github.com/codedellemc/libstorage/blob/master/drivers/storage/ebs/tests/README.md#unitintegration-tests) instructs the user to export their AWS key credentials, but those aren't in the [test config](https://github.com/codedellemc/libstorage/blob/master/drivers/storage/ebs/tests/ebs_test.go#L34-L37) at all.

So when I really think about it, the only reason to have anything in the test file at all is if you want to explicitly give the tests a default other than what the driver already has as a default. It's also worth noting that env vars have priority over the config file, so even if something is defined in the config `[]byte` in the test file, it will be overriden if you export the appropriate variables.

Which brings me to my ultimate conclusion. If you really want to use the OpenStack RC vars by default, it actually does seem like it would be okay to do so directly in the test file. If any of those vars aren't in the environment, they just won't be set by default. However, if a test runner does export env vars that are specific to the driver, those should have precedence over what's defined there in the config string of the test file.

Does that make sense?

I know I've talked myself in a complete circle, but it was a worthwhile journey for me. :)

One other interesting point of reference is what the EFS driver does, which is to build the config [dynamically](https://github.com/codedellemc/libstorage/blob/master/drivers/storage/efs/tests/efs_test.go#L54-L68). If I am reading through it correctly, it first looks for the standard AWS SDK env vars (much like OpenStack RC), and if those aren't present it sets some defaults. And *all* of these would be overriden if a user did something like `export EFS_ACCESSKEY...` or other valid key.